### ISSUE TITLE
Rule: Unnecessary property custom accessor

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#2102]
## Expected Behavior of the rule
```kotlin
val a: Int
  get() = 4
```

should be
```kotlin
val a: Int = 4
```

If a property custom accessor returns a constant value we should not use the custom accessor.

A more complex example:
```kotlin
val a: Asdf
  get() = Asdf("hello", "world")
```

should be
```kotlin
val a: Asdf = Asdf("hello", "world")
```

an example that should not be flagged:
```kotlin
var name: String
val a: Asdf
  get() = Asdf("hello", name)
```

## Context
You can read #2099 to see an example where this rule could be useful.

- What if this property is used as a provider of mutable objects? A function should be better in this case, not a property with a custom accessor.
- What if you don't want to create the property value as soon as you istantiate the class? You can use `lazy`.

**Comment by arturbosch** [comment]
This could be a tricky rule to implement.
Consider
```kotlin
sealed class UType(
    override val name: UName,
    val nullable: Boolean = false,
    override val annotations: Collection<UAnnotation> = emptyList(),
    override val modifiers: Collection<UModifier> = emptyList()
) : UNodeBase(), HasName, HasAnnotations, HasModifiers {

    override val children: Collection<UNode> =
        mergeCollections(annotations, modifiers, listOf(name))
```

When not using `get()`, IntelliJ will complain with `Accessing non-final property name in constructor`.
If you can't change the code to a function, Imo a valid solution would be to use the `get() =` syntax.
Lazy would also work here though
```kotlin
override val children: Collection<UNode> by lazy(LazyThreadSafetyMode.NONE) {
    mergeCollections(annotations, modifiers, listOf(name))
}
```

**Comment by BraisGabin** [comment]
Yeah, the lazy no thread safe is a bit verbose. But in your case that's what you want. You don't want to recreate the children list each time that you access it.

The problem that I see is what if `annotations: Collection<UAnnotation>` change to `annotations: MutableList<UAnnotation>`? Then the only correct way to implement `children` is to use a custom accessor. And it would be flagged by this rule. I'm kind of OK about that because I think that the immutability is the way to go. But I can see that it's controversial.

Anyway, for sure this rule is difficult to implement because you need to check that every value that is used in it is a literal or a `val`. And then assume immutability and that any called function there is pure.

**Comment by cortinico** [comment]
I'd vote for closing this. I belive that such a rule could be extremely controversial as the two notations have implications in terms of Bytecode & Java interop.
